File: app/src/main/java/com/osfans/trime/ime/symbol/CandidateAdapter.java
package com.osfans.trime.ime.symbol;

import android.annotation.SuppressLint;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintSet;
import androidx.recyclerview.widget.RecyclerView;
import com.osfans.trime.core.Rime;
import com.osfans.trime.data.Config;
import com.osfans.trime.databinding.LiquidKeyItemBinding;
import java.util.ArrayList;
import java.util.List;

// 显示长度不固定，字体大小正常的内容。用于类型 CANDIDATE, VAR_LENGTH
public class CandidateAdapter extends RecyclerView.Adapter<CandidateAdapter.ViewHolder> {
  @NonNull private final List<Rime.RimeCandidate> mCandidates = new ArrayList<>();

  private final Config theme;

  public CandidateAdapter(@NonNull Config theme) {
    this.theme = theme;
  }

  enum CommentPosition {
    UNKNOWN,
    TOP,
    BOTTOM,
    RIGHT;
  }

  public void updateCandidates(List<Rime.RimeCandidate> candidates) {
    mCandidates.clear();
    mCandidates.addAll(candidates);
  }

  @Override
  public int getItemCount() {
    return mCandidates.size();
  }

  @NonNull
  @Override
  public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    final LiquidKeyItemBinding binding =
        LiquidKeyItemBinding.inflate(LayoutInflater.from(parent.getContext()));
    final boolean isCommentHidden = Rime.get_option("_hide_comment");
    if (!isCommentHidden) {
      final ConstraintSet set = new ConstraintSet();
      set.clone(binding.getRoot());
      final CommentPosition commentPosition =
          CommentPosition.values()[theme.getInt("comment_position")];
      switch (commentPosition) {
        case BOTTOM:
          set.centerHorizontally(binding.comment.getId(), ConstraintSet.PARENT_ID);
          set.centerHorizontally(binding.candidate.getId(), ConstraintSet.PARENT_ID);
          set.connect(
              binding.comment.getId(),
              ConstraintSet.TOP,
              binding.candidate.getId(),
              ConstraintSet.BOTTOM);
          set.connect(
              binding.comment.getId(),
              ConstraintSet.BOTTOM,
              ConstraintSet.PARENT_ID,
              ConstraintSet.BOTTOM);
          break;
        case TOP:
          set.centerHorizontally(binding.comment.getId(), ConstraintSet.PARENT_ID);
          set.centerHorizontally(binding.candidate.getId(), ConstraintSet.PARENT_ID);
          set.connect(
              binding.comment.getId(),
              ConstraintSet.BOTTOM,
              binding.candidate.getId(),
              ConstraintSet.TOP);
          set.connect(
              binding.comment.getId(),
              ConstraintSet.TOP,
              ConstraintSet.PARENT_ID,
              ConstraintSet.TOP);
          break;
        case RIGHT:
        case UNKNOWN:
        default:
          break;
      }
      set.applyTo(binding.getRoot());
    }
    binding.comment.setVisibility(View.GONE);
    return new ViewHolder(binding);
  }

  static class ViewHolder extends RecyclerView.ViewHolder {
    final TextView candidate, comment;

    public ViewHolder(@NonNull LiquidKeyItemBinding binding) {
      super(binding.getRoot());
      candidate = binding.candidate;
      comment = binding.comment;
    }
  }

  @SuppressLint("ClickableViewAccessibility")
  @Override
  public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
    final Rime.RimeCandidate candidate = mCandidates.get(position);

    final Typeface candidateFont = theme.getFont("candidate_font");
    final Typeface commentFont = theme.getFont("comment_font");
    holder.candidate.setTypeface(candidateFont);
    holder.comment.setTypeface(commentFont);

    holder.candidate.setText(candidate.text);
    if (holder.comment.getVisibility() == View.VISIBLE) {
      holder.comment.setText(candidate.comment);
    }

    final float candidateSize = theme.getFloat("candidate_text_size");
    final float commentSize = theme.getFloat("comment_text_size");
    if (candidateSize > 0) holder.candidate.setTextSize(candidateSize);
    if (commentSize > 0) holder.comment.setTextSize(commentSize);

    final int candidateColor = theme.getColor("candidate_text_color");
    final int commentColor = theme.getColor("comment_text_color");
    holder.candidate.setTextColor(candidateColor);
    holder.comment.setTextColor(commentColor);

    //  点击前后必须使用相同类型的背景，或者全部为背景图，或者都为背景色
    // 如果直接使用background，会造成滚动时部分内容的背景填充错误的问题
    final Drawable background =
        theme.getDrawable("key_back_color", "key_border", "key_border_color", "round_corner", null);
    if (background != null) holder.itemView.setBackground(background);

    // 如果设置了回调，则设置点击事件
    if (listener != null) {
      holder.itemView.setOnClickListener(view -> listener.onCandidateClick(position));
    }

    // 点击时产生背景变色效果
    holder.itemView.setOnTouchListener(
        (view, motionEvent) -> {
          final int hilited = theme.getColor("hilited_candidate_text_color");
          if (motionEvent.getAction() == MotionEvent.ACTION_DOWN) {
            holder.candidate.setTextColor(hilited);
          }
          return false;
        });
  }

  /** 添加 Listener 回调 * */
  public interface Listener {
    void onCandidateClick(int position);
  }

  private Listener listener;

  public void setListener(@NonNull Listener listener) {
    this.listener = listener;
  }
}

File: app/src/main/java/com/osfans/trime/ime/symbol/CandidateAdapter.kt
// SPDX-FileCopyrightText: 2015 - 2024 Rime community
//
// SPDX-License-Identifier: GPL-3.0-or-later

package com.osfans.trime.ime.symbol

import android.content.Context
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.StateListDrawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.view.updateLayoutParams
import androidx.recyclerview.widget.RecyclerView
import com.chad.library.adapter4.BaseQuickAdapter
import com.osfans.trime.core.CandidateListItem
import com.osfans.trime.core.Rime
import com.osfans.trime.data.theme.ColorManager
import com.osfans.trime.data.theme.FontManager
import com.osfans.trime.data.theme.Theme
import com.osfans.trime.databinding.LiquidEntryViewBinding
import splitties.views.dsl.constraintlayout.above
import splitties.views.dsl.constraintlayout.below
import splitties.views.dsl.constraintlayout.bottomOfParent
import splitties.views.dsl.constraintlayout.centerHorizontally
import splitties.views.dsl.constraintlayout.topOfParent

// 显示长度不固定，字体大小正常的内容。用于类型 CANDIDATE, VAR_LENGTH
class CandidateAdapter(private val theme: Theme) : BaseQuickAdapter<CandidateListItem, CandidateAdapter.ViewHolder>() {
    enum class CommentPosition {
        UNKNOWN,
        TOP,
        BOTTOM,
        RIGHT,
    }

    private val mCandidateTextSize = theme.generalStyle.candidateTextSize.toFloat().coerceAtLeast(1f)
    private val mCandidateFont = FontManager.getTypeface("candidate_font")
    private val mCandidateTextColor = ColorManager.getColor("candidate_text_color")
    private val mHilitedCandidateBackColor = ColorManager.getColor("hilited_candidate_back_color")
    private val mCommentPosition = theme.generalStyle.commentPosition
    private val mCommentTextSize = theme.generalStyle.commentTextSize.toFloat().coerceAtLeast(1f)
    private val mCommentFont = FontManager.getTypeface("comment_font")
    private val mCommentTextColor = ColorManager.getColor("comment_text_color")

    override fun onCreateViewHolder(
        context: Context,
        parent: ViewGroup,
        viewType: Int,
    ): ViewHolder {
        val binding = LiquidEntryViewBinding.inflate(LayoutInflater.from(parent.context))
        binding.root.background =
            StateListDrawable().apply {
                addState(
                    intArrayOf(),
                    ColorManager.getDrawable(
                        context = context,
                        key = "key_back_color",
                        border = theme.generalStyle.candidateBorder,
                        borderColorKey = "key_border_color",
                        roundCorner = theme.generalStyle.roundCorner,
                    ),
                )
                mHilitedCandidateBackColor?.let {
                    addState(intArrayOf(android.R.attr.state_pressed), ColorDrawable(it))
                }
            }
        binding.candidate.apply {
            textSize = mCandidateTextSize
            typeface = mCandidateFont
            mCandidateTextColor?.let { setTextColor(it) }
        }
        val isCommentHidden = Rime.getRimeOption("_hide_comment")
        if (isCommentHidden) return ViewHolder(binding)

        binding.comment.apply {
            visibility = View.GONE
            textSize = mCommentTextSize
            typeface = mCommentFont
            mCommentTextColor?.let { setTextColor(it) }
        }
        val candidate = binding.candidate
        val comment = binding.comment
        when (mCommentPosition) {
            CommentPosition.BOTTOM -> {
                candidate.updateLayoutParams<ConstraintLayout.LayoutParams> {
                    centerHorizontally()
                }
                comment.updateLayoutParams<ConstraintLayout.LayoutParams> {
                    below(candidate)
                    centerHorizontally()
                    bottomOfParent()
                }
            }

            CommentPosition.TOP -> {
                candidate.updateLayoutParams<ConstraintLayout.LayoutParams> {
                    centerHorizontally()
                }
                comment.updateLayoutParams<ConstraintLayout.LayoutParams> {
                    topOfParent()
                    above(candidate)
                    centerHorizontally()
                }
            }

            CommentPosition.RIGHT, CommentPosition.UNKNOWN -> {}
        }
        return ViewHolder(binding)
    }

    class ViewHolder(binding: LiquidEntryViewBinding) : RecyclerView.ViewHolder(binding.root) {
        val candidate: TextView = binding.candidate
        val comment: TextView = binding.comment
    }

    override fun onBindViewHolder(
        holder: ViewHolder,
        position: Int,
        item: CandidateListItem?,
    ) {
        item?.run {
            holder.candidate.text = text
            holder.comment.text = comment
        }
    }
}
